src/App.tsx

```diff
@@ -15,48 +16,110 @@ export const goodsFromServer = [
   'Garlic',
 ];
 
+enum SortBy {
+  Alphabet = 'alphabet',
+  Length = 'length',
+}
+
+type StateVar = {
+  sortBy: string,
+  isReverse: boolean,
+};
+
+type Func = (goods: string[], { sortBy, isReverse }: StateVar) => string[];
+
+const getPrepareGoods: Func = (goods, { sortBy, isReverse }) => {
+  const prepareGoods = [...goods];
+
+  if (sortBy) {
+    prepareGoods.sort((goodA, goodB) => {
+      switch (sortBy) {
+        case SortBy.Alphabet:
+          return goodA.localeCompare(goodB);
+        case SortBy.Length:
+          return goodA.length - goodB.length;
+        default:
+          return 0;
+      }
+    });
+  }
+
+  if (isReverse) {
+    prepareGoods.reverse();
+  }
+
+  return prepareGoods;
+};
+
 export const App: React.FC = () => {
+  const [sortBy, setSortBy] = useState('');
+  const [isReverse, setIsReverse] = useState(false);
+
+  const goods = getPrepareGoods(goodsFromServer, { sortBy, isReverse });
+
+  const GoodItems = () => (
+    <ul>
+      {goods.map(good => (
+        <li key={good} data-cy="Good">{good}</li>
+      ))}
+    </ul>
+  );
+
+  const handleSort = (sort: string) => () => setSortBy(sort);
+  const handleReverse = () => setIsReverse(!isReverse);
+
+  const reset = () => {
+    setSortBy('');
+    setIsReverse(false);
+  };
+
   return (
     <div className="section content">
       <div className="buttons">
         <button
+          onClick={handleSort(SortBy.Alphabet)}
           type="button"
-          className="button is-info is-light"
+          className={
+            cn('button is-info', { 'is-light': sortBy !== 'alphabet' })
+          }
         >
           Sort alphabetically
         </button>
 
         <button
+          onClick={handleSort(SortBy.Length)}
           type="button"
-          className="button is-success is-light"
+          className={
+            cn('button is-success', { 'is-light': sortBy !== 'length' })
+          }
         >
           Sort by length
         </button>
 
         <button
+          onClick={handleReverse}
           type="button"
-          className="button is-warning is-light"
+          className={
+            cn('button is-warning', { 'is-light': !isReverse })
+          }
         >
           Reverse
         </button>
 
-        <button
-          type="button"
-          className="button is-danger is-light"
-        >
-          Reset
-        </button>
+        {(isReverse || sortBy)
+          && (
+            <button
+              onClick={reset}
+              type="button"
+              className="button is-danger is-light"
+            >
+              Reset
+            </button>
+          )}
       </div>
 
       <ul>
-        <ul>
-          <li data-cy="Good">Dumplings</li>
-          <li data-cy="Good">Carrot</li>
-          <li data-cy="Good">Eggs</li>
-          <li data-cy="Good">Ice cream</li>
-          <li data-cy="Good">Apple</li>
-          <li data-cy="Good">...</li>
-        </ul>
+        <GoodItems />
```

Comment:
Render < ul > (the best way doesn't use small components inside other components, or move them to another file). ( not critical, recommendation)